src/lib.rs

```diff
@@ -2184,4 +2200,35 @@ mod tests {
         assert!(kresolver.ksymresolver.is_some());
         assert!(kresolver.kernelresolver.is_none());
     }
+
+    #[test]
+    fn load_gsym_resolver() {
+        let args: Vec<String> = env::args().collect();
+        let bin_name = &args[0];
+        let test_gsym = Path::new(bin_name)
+            .parent()
+            .unwrap()
+            .parent()
+            .unwrap()
+            .parent()
+            .unwrap()
+            .parent()
+            .unwrap()
+            .join("data")
+            .join("test.gsym");
+        let base: u64 = 0x77a7000; // pickup randomly.
```

Comment:
If we don't compile samples from the source every time, adding linker scripts doesn't add value.  Do you want to compile samples every time?

Reply:
It adds value even if we don't. Another toolchain or toolchain version could place stuff at different locations, right? So now, if we decide to change something on the file, we need to recreate the blob, but don't have the very same version of the original toolchain available, we end up yak shaving test failures of potentially all tests relying on the file in one form or another, being forced to fix each one up. So I don't think we should add an unstable binary blob like that.

Reply:
First, I am sorry for replying to these comments late.  I didn't see these comments for some reason until I cleaned the cache.

Let me clarify a little bit. `base` is an address to relocate the object to.  It simulates that an object (binary) is relocated to this address, not being loaded at address 0x0.  I can replace it with a random number to make it even better.

`0x400596` is the hard-coded address.

This test case here is to ensure the code works correctly and consistently on a specific sample.   If we want to explore potential bugs, it should test against bigger samples separately.   For example, in our other test cases, they test against the binary generated by `cargo test`.   However, it will need additional tools to generate GSYM files.  It should work on a specific test environment with a separate test suite.

Otherwise, I don't see any reason we need to rebuild this sample.  If you want to test another use case, add a new sample instead of changing existing samples.  Changing samples may allow regressions without notice.  For example, some bugs are caught only by one specific sample.  Once changing the sample, the bugs may be reintroduced, not caught.  I am not saying this sample is any special.  I am saying we need to add new samples if we want to test something that this sample doesn't catch.

Reply:
> Let me clarify a little bit. base is an address to relocate the object to. It simulates that an object (binary) is relocated to this address, not being loaded at address 0x0.
> 
> 0x400596 is the hard-coded address.

Makes sense, though I am not so much interested in this specific instance and its purpose, as I am making a general remark. Here https://github.com/libbpf/blazesym/pull/22#discussion_r1043899675 we certainly are assuming a symbol at a very specific address, right?

> This test case here is to ensure the code works correctly and consistently on a specific sample. If we want to explore potential bugs caught by different samples, it should test against bigger samples separately.

I am not disagreeing with that statement in principle, but I think you draw very different conclusions from it than I do. For you, it seems to be an excuse to check in whatever binary comes along that fits the purpose, hard code a bunch of numbers matching this very binary in a test, and call it a day.

To me, the consistency does not come from checking in a binary and hard wiring a bunch of address from it into the code base, it comes from generating this binary according to a recipe that makes the specific properties that the generated artifact has to carry (because they are what is being tested!) explicit. You want DWARF v4 format? You tell the compiler to build the binary with it. You want a symbol at a specific address, you tell the linker to place it there.

Then, every test failure is either a bug in the test itself, `blazesym`, or the compiler, and needs to be looked at consequently. It does not matter what environment.

> For example, in our other test cases, they test against the binary generated by cargo test.

Yes, a questionable practice in itself, as vividly illustrated by the bunch of work I had to do to fix up those tests here: https://github.com/libbpf/blazesym/pull/43

> However, it will need additional tools to generate GSYM files.

Sure, those will need to be installed.

> Otherwise, I don't see any reason we need to rebuild this sample. If you want to test another use case, add a new sample instead of changing existing samples. Changing samples may allow regressions without notice. For example, some bugs are caught only by one specific sample. Once changing the sample, the bugs may be reintroduced, not caught. I am not saying this sample is any special. I am saying we need to add new samples if we want to test something that this sample doesn't catch.

We can certainly argue *when* to rebuild test binaries and whether there is a case to be made for checking them in, but please stop advocating in favor of having a bunch of immutable black boxes being depended on by our tests, with the assumption of never changing them moving forward for reasons of regression testing. This is not regression testing, it is metaphorically speaking, throwing a bunch of stuff against the wall and seeing what sticks. If you want a regression test then first you have to define the regression being tested.

What what you are saying amounts to is: we have a bunch of read-only tests and a bunch of read-only test binaries. We cannot change either. If a test starts failing because, say, we made a change in `blazesym`, all we basically can do is remove the now failing test. We don't know what is in the binary, so for all we know the changes to `blazesym` may be correct and the test was passing earlier because it relied on a toolchain bug present in the checked in binary, generated years ago.

I am not going to debug something like that.

Now, there may be certain limits to what can be specified in a recipe, which *potentially* *may* warrant checking-in a binary blob generated in a one-off fashion based on a known toolchain version etc., but we are far from that here.

What is being done here is in my opinion and to put it lightly, sloppy engineering.

Reply:
@danielocfb I'm not very familiar with linker scripts, so sorry if it's a stupid question. But does it allow easily to put a specific function at specific offset relative to base load address? I thought that with linker scripts it's easy (after you familiarize with yet another custom DSL, but that aside) to specify stuff like fixed address at which entire ELF section or segment is loaded stuff like that. But hard-coding that fibonacci() function has to be loaded at offset 123 -- how easy is that?

But even if easy, would it make sense to do it for any custom test case? Seems pretty tedious, tbh.

From my perspective, test.gsym is not really a black box. You can dump it using Clang's standard tooling (llvm-gsymutil), e.g., I'm getting: 
```
$ llvm-gsymutil test.gsym
Header:
  Magic        = 0x4753594d
  Version      = 0x0001
  AddrOffSize  = 0x02
  UUIDSize     = 0x14
  BaseAddress  = 0x0000000000400000
  NumAddresses = 0x0000000d
  StrtabOffset = 0x00000094
  StrtabSize   = 0x000000fa
  UUID         = 46a7f0cc94901d5e329162b2166105a8736cd0ed

Address Table:
INDEX  OFFSET16 (ADDRESS)
====== ===============================
...
FunctionInfo @ 0x00000248: [0x00000000004005db - 0x0000000000400625) "main"
LineTable:
  0x00000000004005db /home/kuifeng/progm/blazesym/data/gsym-example.c:23
  0x00000000004005dc /home/kuifeng/progm/blazesym/data/gsym-example.c:24
  0x00000000004005dc /home/kuifeng/progm/blazesym/data/gsym-example.c:26
  0x00000000004005dc /home/kuifeng/progm/blazesym/data/gsym-example.c:11
  0x00000000004005dc /home/kuifeng/progm/blazesym/data/gsym-example.c:12
  0x00000000004005f7 /home/kuifeng/progm/blazesym/data/gsym-example.c:27
  0x00000000004005fc /home/kuifeng/progm/blazesym/data/gsym-example.c:28
  0x0000000000400616 /home/kuifeng/progm/blazesym/data/gsym-example.c:27
  0x000000000040061e /home/kuifeng/progm/blazesym/data/gsym-example.c:30
  0x000000000040061e /home/kuifeng/progm/blazesym/data/gsym-example.c:31
InlineInfo:
[0x00000000004005db - 0x0000000000400625) main
  [0x00000000004005dc - 0x00000000004005f7) factorial_inline_wrapper called from /home/kuifeng/progm/blazesym/data/gsym-example.c:26
FunctionInfo @ 0x00000294: [0x0000000000400630 - 0x0000000000400695) "__libc_csu_init"
FunctionInfo @ 0x000002a4: [0x00000000004006a0 - 0x00000000004006a5) "__libc_csu_fini"
FunctionInfo @ 0x000002b4: [0x00000000004006a8 - 0x00000000004006b5) "_fini"
```

Having a standard tool that allows to introspect input test data is great and makes it not that much of a black box, IMO. Similar for (small) checked-in ELF files. Unless we have reasonable expectations that we'll have to periodically re-generate input data for some reason, it feels adequate enough to have original source code, instructions how to turn that source code into a binary artifact (gsym file in this case), and then write tests with the knowledge of specific absolute addresses/offsets/names/etc.

I'm not sure how unorthodox it seems, but the thought of having to write linker script for every example is pretty daunting, tbh. We do check in stable binary artifacts, we have source code to make a connection between C source code and binary format (plus tools to dump them in human-readable format, llvm-gsymutil in this case), that seems adequate for lots of small tests that test some particular aspect of resolve functionality? WDYT?

Just one thought (and I'm not even convinced it is an improvement per se, but thought I'll bring it up) to improve tests "comprehension" a bit. Maybe we should hard-code explicitly-named constants. E.g., in this case we'd have a) base load address (0x400000), which is unlikely to change, but if we are concerned we can have small common linker script that puts sections at 0x400000 forever and b) function offset and function size. The latter are both kind of unstable by nature, different compiler versions will end up generating different assembly code, potentially. I'm not sure there is anything that can be done, short of writing embedded assembly (horror).

As I said, not convinced it's an improvement, but assuming we'll have lots of small input files, having base load as a constant will make it easier to update it if some sort of linker/compiler default changes in the future. Not sure if anything can be done about function offset and size (and especially source code location to function offset mapping, that seems extremely unstable, right?)

Anyway, my 2 cents, hope you guys don't mind me intruding a bit.

Reply:
You can put functions of interest into a separate section and place the section where you want it to be.

Anyway, given that I am seemingly the only one feeling strongly about this kind of work, I'll bend to the majority opinion and won't bring it up again.

Reply:
>To me, the consistency does not come from checking in a binary and hard wiring a bunch of address from it into the code base, it comes from generating this binary according to a recipe that makes the specific properties that the generated artifact has to carry (because they are what is being tested!) explicit. You want DWARF v4 format? You tell the compiler to build the binary with it. You want a symbol at a specific address, you tell the linker to place it there.
>
>...
>
>What what you are saying amounts to is: we have a bunch of read-only tests and a bunch of read-only test binaries. We cannot change either. If a test starts failing because, say, we made a change in blazesym, all we basically can do is remove the now failing test. We don't know what is in the binary, so for all we know the changes to blazesym may be correct and the test was passing earlier because it relied on a toolchain bug present in the checked in binary, generated years ago.
>
>I am not going to debug something like that.

I personally agree pretty strongly with Daniel here. Checking in binaries is, in my opinion, not a great way to set up an engineering pipeline for testing. IMO we should instead be formally defining testcases as something like a tuple of (code, optional linker script) x (toolchain, expected output). I think this makes it easier to debug, but more importantly, you don't have to manually compile and check-in a binary for every testcase for every single toolchain (gcc, clang, DWARF4, DWARF5) you want to support. That doesn't really scale, and it will raise the barrier to entry for quickly being able to test new toolchains. I agree that each individual case maybe isn't that bad, but why not just pay the tax up front and have something that's easily extensible (and easier to reason about) rather than checking in a black box binary? The linker version script and code (which is compiled by the testing pipeline) can be shared for each version of the toolchain it's testing.

@anakryiko Agreed that linker scripts are kind of arcane, but they do work pretty well and aren't that complicated once you're familiar with them.

>But hard-coding that fibonacci() function has to be loaded at offset 123 -- how easy is that?

I think that's pretty easy. Can't you just do:

```
SECTIONS
{
fibonacci = 0x123;
}
```
or if you want to load it at offset 0x123 from the .text section, I believe you could do:
```
SECTIONS
{
    .text :
        {
             fibonacci = ABSOLUTE(.) + 0x123;
             *(.text)
        }
}
```
Might have to fiddle with it, but I'm almost certain both of the things you mentioned are possible with very minimal effort. That kind of stuff is linker script 101, and I don't expect that to be the heavy lift. The heavy lift will probably be setting up a testing pipeline that automates all of this, but IMO that's the right place to invest more time because it makes it easier to add testcases for new toolchains in the future.

>But even if easy, would it make sense to do it for any custom test case? Seems pretty tedious, tbh.

Hard to say, it's admittedly subjective. But IMO it's a good long-term investment to set up a testing pipeline that's programmatic, and which allows you to easily test blazesym on different toolchains.

>I'm not sure how unorthodox it seems, but the thought of having to write linker script for every example is pretty daunting, tbh. We do check in stable binary artifacts, we have source code to make a connection between C source code and binary format (plus tools to dump them in human-readable format, llvm-gsymutil in this case), that seems adequate for lots of small tests that test some particular aspect of resolve functionality? WDYT?

I think perhaps part of the disconnect here is that a symbolizer is a bit of a special type of tool when it comes to testing. We're not really trying to test that a version of clang and gcc both put a variable at the expected offset for a linker script, what we're really testing is "does this tool parse GSYM / ELF / DWARF / etc correctly?" IMO though, there is a lot of value in setting up the pipeline this way because e.g. a new compiler could output ELF / DWARF / GSYM in a different manner than what gsym is expecting, and I think we want to catch that. Otherwise we'd have to manually compile all of these testcases for every single toolchain we want to support, and then once everything works, check all of the binaries into the codebase. That seems super inconvenient.

Reply:
> > I'm not sure how unorthodox it seems, but the thought of having to write linker script for every example is pretty daunting, tbh. We do check in stable binary artifacts, we have source code to make a connection between C source code and binary format (plus tools to dump them in human-readable format, llvm-gsymutil in this case), that seems adequate for lots of small tests that test some particular aspect of resolve functionality? WDYT?
> 
> I think perhaps part of the disconnect here is that a symbolizer is a bit of a special type of tool when it comes to testing. We're not really trying to test that a version of clang and gcc both put a variable at the expected offset for a linker script, what we're really testing is "does Blazesym parse GSYM / ELF / DWARF / etc correctly in this scenario?", so I see why checking in a binary would be a reasonable suggestion.

Exactly. We had some more offline discussion with @danielocfb which weren't captured here. While in this specific case where we just have one function and we just trying to symbolize from any address within it to its name it's possible to just specify everything rigidly, the very next immediate step is to do the same for the same function, but now try to map some specific offset of an assembly instruction within fibonacci to line info (file + line number). This you can't nicely control with linker script. Another immediate use case would be inlined function within non-inlined function. Again, you can't control where and how exactly compiler intersperses assembly instructions within fibonacci(). And we can continue scenarios like this, gradually complicating them. And the only things that are controllable is base load address and function initial offsets. But gsym and dwarf data has way more and we need to test all that.

I don't see how we can test it differently from generating small binary (ELF, DWARF dump, GSYM dump, whatever makes sense) and using that as fixed check-in input test case. At that point there is little benefit to knowing exactly that base load address is 0x400000 and fibonacci is at offset 0x123 from it, if we really need to be sure that offset 0x555 within fibonacci maps to line number 43 (as a random example, of course).

So with that said, I don't see the point of fixing 2 variables when we can't control all the other at least as important ones.

> IMO though, there is a lot of value in programatically setting up the "scenario" to make it easier to onboard new toolchains.

So I sympathize in principle about reproducible test cases. But taking the above into account, I think small checked in binaries are the only reasonable (and consistent) way to go about this.


As for scalability. Whoever adds new test case generates one-two-three files per case, that seems scalable enough. We don't need to re-generate those binaries all the time, right?

Reply:
> I think that's pretty easy. Can't you just do:
> 
> SECTIONS
> {
> fibonacci = 0x123;
> }

It changes the value of the symbol, but only "the value of the symbol".  It doesn't move the code to the address you assigned.

The syntax of linker scripts may be not complicated.  It is very straightforward in some Linux distros, but not all.  Toolchains and platforms do a lot that would change the result.  For example, libc may expect you to set the script in a specific way, and it varies by version and distro.  Toolchains may also add additional sections to the output.  Not to mention that we will test inline functions later.

Reply:
> While in this specific case where we just have one function and we just trying to symbolize from any address within it to its name it's possible to just specify everything rigidly, the very next immediate step is to do the same for the same function, but now try to map some specific offset of an assembly instruction within fibonacci to line info (file + line number). This you can't nicely control with linker script. Another immediate use case would be inlined function within non-inlined function. Again, you can't control where and how exactly compiler intersperses assembly instructions within fibonacci(). And we can continue scenarios like this, gradually complicating them. And the only things that are controllable is base load address and function initial offsets. But gsym and dwarf data has way more and we need to test all that.

Sure, we may want to test that. But that does not have to happen using this kind of test. Any sufficiently complex system is tested in layers (refer to [today's trending post](https://duckduckgo.com/?q=software+testing+pyramid&ia=web) of the software testing pyramid). Having the presence of one approach to testing rule out that of another is far from the conclusion I would draw. On the contrary, it goes against the desire to achieve high test coverage.

To make it concrete, we get additional exposure to toolchain quirks and similar, that we may very well interested in understanding, because these are properties that real life binaries may exhibit as well, by testing with artifacts produced on different toolchains. We don't really get that by dumping a bunch of one-off created blobs into the code base.

And as I said, this is not the only approach to testing. For the specific example you brought up, the code path may easily be coverable by a unit test. Or we could just specify a range of offsets that could possibly be present. These approaches combine. Of course we have to think a bit about it.

To digress a bit, Rust is very amenable to unit testing and one can easily help test at different layers. You have a function in there you can pretty much test it. Yes, it may require some co-design between library itself and the set of intended tests (some clean abstractions are also helpful, e.g., to easily capture outputs or similar). I'll argue that no sufficiently complex system achieves good test coverage and reasonable test runtime by just fudging a set of end-to-end tests around it and calling it a day.

That is not trivially possible in C and so rarely any projects do it properly. You have candidates such as [`CUnit`](https://cunit.sourceforge.net/) which go into that direction, but let's face it, it's horrible to work with owing to deficiencies in language/tooling surrounding it. So in the end you'd end up building your own thing and at best a minority does that.

But back on topic, I never explicitly ruled out checking in a binary:

> Now, there may be certain limits to what can be specified in a recipe, which potentially may warrant checking-in a binary blob generated in a one-off fashion based on a known toolchain version etc., but we are far from that here.

What I am saying is that there is no reason to make it the default approach.

Reply:
> What I am saying is that there is no reason to make it the default approach.

Sure, I didn't know we are picking just one way of testing here. I totally agree that we should have unit tests. But having end-to-end tests from binary input data that users will be providing in practice to expected output (func name, file name, line num, etc) is a must, IMO. I don't see an alternative for this end-to-end testing besides checking in binary inputs (and given we are checking in binaries, I don't see the point in linker script that fixes some aspects of that binary either), but unit-testing is great at catching and pinpointing to issues way more precisely and easily (debugging end-to-end test is certainly less pleasant).

I'm not sure, but I think we are in agreement, in the end? But let me know if not.

Reply:
> I'm not sure, but I think we are in agreement, in the end? But let me know if not.

Not really, no.
- I have made the case that we get benefits from having a full build from source kind of thing for (some) tests and I haven't seen that premise questioned.
- Second, I think we are also in agreement that we can make it work for some tests, because the relevant inputs can be meaningfully controlled. In fact, that has been started as part of this pull request with the linker script added.
- Third, we seem to agree that this one approach does not have to cover all conceivable properties we want to test moving forward, because other ways of testing are available (including the option of falling back to using a checked-in binary). It is, however, very much relevant to the direction we take, because once we rule that kind of testing out here, I can pretty much assure you it's not coming back. Combine that with 1/ and I think we objectively loose out on coverage.

So, given these three beliefs, I wonder, why not go all the way *here and now*? The only reasons I've really seen against that were that
- we'd have to spend time on understanding the tooling available to us (e.g., learning the workings of linker scripts) and
- we need to do it to begin with, which, again, seems to boil down to the time component.

While I am not arguing that engineering effort is irrelevant, we probably could have put the entire "framework" in place three times over in the time we argued here -- meaning it wasn't all that much work to begin with. And even if we consider the development moving forward, if there is a strong case to be made that controlling all relevant variables is too expensive (on any axis, including time/effort), we *always* have the option of falling back to checking in a binary (though again, while that is an option, it is *not* the *only* option).

Reply:
Linker scripts are reliable in a wild environment without specific toolchains.  Even a simple link script here; I also found I have to adjust it to make it work not only for my development machine but also my laptop because the difference of libc put some restrictions on linker scripts.  In some platforms, only the linker scripts in a specific way would work.  If every test case has to experience these, it is too difficult to be used in our project.

However, if you believe it wasn't much work, you should invest time to try it and find a solution that works for our cases.

Reply:
> So, given these three beliefs, I wonder, why not go all the way _here and now_? The only reasons I've really seen against that were that
> 
>     * we'd have to spend time on understanding the tooling available to us (e.g., learning the workings of linker scripts) and
> 
>     * we need to do it to begin with, which, again, seems to boil down to the time component.
> 

Hmm.. I thought the main argument was this (from https://github.com/libbpf/blazesym/pull/22#discussion_r1061989389):

> While in this specific case where we just have one function and we just trying to symbolize from any address within it to its name it's possible to just specify everything rigidly, the very next immediate step is to do the same for the same function, but now try to map some specific offset of an assembly instruction within fibonacci to line info (file + line number). This you can't nicely control with linker script. Another immediate use case would be inlined function within non-inlined function. Again, you can't control where and how exactly compiler intersperses assembly instructions within fibonacci(). And we can continue scenarios like this, gradually complicating them. And the only things that are controllable is base load address and function initial offsets. But gsym and dwarf data has way more and we need to test all that.
> 
> I don't see how we can test it differently from generating small binary (ELF, DWARF dump, GSYM dump, whatever makes sense) and using that as fixed check-in input test case. At that point there is little benefit to knowing exactly that base load address is 0x400000 and fibonacci is at offset 0x123 from it, if we really need to be sure that offset 0x555 within fibonacci maps to line number 43 (as a random example, of course).


But as I said, I just don't see how tests like this can be done some other way, but I'm happy to be shown otherwise. I see this PR was merged, so do you intend to make these improvements you mentioned in separate PR?

In any case, I think we should be open to have binary inputs for end-to-end testing (as was mentioned above, it seems like llvm-symbolizer took similar approach, so it must be working for them well?), but as you said, if it's faster to just do the approach you are proposing, let's go with it.